src/tr2/game/lara/control.c

```diff
@@ -746,66 +739,47 @@ void Lara_Animate(ITEM *const item)
     }
 
-    if (anim->num_commands > 0) {
-        const int16_t *cmd_ptr = Anim_GetCommand(anim->command_idx);
-        for (int32_t i = 0; i < anim->num_commands; i++) {
-            const int16_t cmd = *cmd_ptr++;
-
-            switch (cmd) {
-            case AC_MOVE_ORIGIN:
-                cmd_ptr += 3;
-                break;
+    for (int32_t i = 0; i < anim->num_commands; i++) {
+        const ANIM_COMMAND *const command = &anim->commands[i];
 
-            case AC_JUMP_VELOCITY:
-                cmd_ptr += 2;
-                break;
-
-            case AC_SOUND_FX: {
-                const int32_t frame = cmd_ptr[0];
-                const SOUND_EFFECT_ID sound_id =
-                    ANIM_CMD_PARAM_BITS(cmd_ptr[1]);
-                const ANIM_COMMAND_ENVIRONMENT type =
-                    ANIM_CMD_ENVIRONMENT_BITS(cmd_ptr[1]);
-                cmd_ptr += 2;
-
-                if (item->frame_num != frame) {
-                    break;
-                }
-
-                if (type == ACE_ALL
-                    || (type == ACE_LAND
-                        && (g_Lara.water_surface_dist >= 0
-                            || g_Lara.water_surface_dist == NO_HEIGHT))
-                    || (type == ACE_WATER && g_Lara.water_surface_dist < 0
-                        && g_Lara.water_surface_dist != NO_HEIGHT)) {
-                    Sound_Effect(sound_id, &item->pos, SPM_ALWAYS);
-                }
+        switch (command->type) {
+        case AC_SOUND_FX: {
+            const COMMAND_EFFECT_DATA *const data =
+                (COMMAND_EFFECT_DATA *)command->data;
+            if (item->frame_num != data->frame_num) {
                 break;
             }
 
-            case AC_EFFECT:
-                const int32_t frame = cmd_ptr[0];
-                const int32_t action_id = ANIM_CMD_PARAM_BITS(cmd_ptr[1]);
-                const ANIM_COMMAND_ENVIRONMENT type =
-                    ANIM_CMD_ENVIRONMENT_BITS(cmd_ptr[1]);
-                cmd_ptr += 2;
-
-                if (item->frame_num != frame) {
-                    break;
-                }
-
-                if (type == ACE_ALL
-                    || (type == ACE_LAND
-                        && (g_Lara.water_surface_dist >= 0
-                            || g_Lara.water_surface_dist == NO_HEIGHT))
-                    || (type == ACE_WATER && g_Lara.water_surface_dist < 0)) {
-                    ItemAction_Run(action_id, item);
-                }
+            const ANIM_COMMAND_ENVIRONMENT type = data->environment;
+            if (type == ACE_ALL
+                || (type == ACE_LAND
+                    && (g_Lara.water_surface_dist >= 0
+                        || g_Lara.water_surface_dist == NO_HEIGHT))
+                || (type == ACE_WATER && g_Lara.water_surface_dist < 0
+                    && g_Lara.water_surface_dist != NO_HEIGHT)) {
+                Sound_Effect(data->effect_num, &item->pos, SPM_ALWAYS);
```

Comment:
Offtop – can we use `SPM_UNDERWATER` to simplify the checks above?
Or alternatively, introduce a getter to retrieve Lara (or an `ITEM`) current environment.

Reply:
Yes, I think this can be made much simpler, similar to `Item_PlayAnimSFX` in TR1. I think I'll aim to make a common function for this in both games to tidy it all up (and this will help get us near a common `Item/Lara_Animate`). But I think tackling it on its own would be best rather than changing logic here, WDYT?

Reply:
Yes definitely a separate effort so as not to muddle up this PR.